relayer/src/chain/cosmos/wait.rs

```diff
@@ -1,114 +1,117 @@
+/// Given a vector of `TxHash` elements,
 /// each including a transaction response hash for one or more messages, periodically queries the chain
 /// with the transaction hashes to get the list of IbcEvents included in those transactions.
 pub async fn wait_for_block_commits(
     chain_id: &ChainId,
     rpc_client: &HttpClient,
     rpc_address: &Url,
     rpc_timeout: &Duration,
-    tx_sync_results: &mut [TxSyncResult],
-) -> Result<(), Error> {
-    let start_time = Instant::now();
-
-    let hashes = tx_sync_results
-        .iter()
-        .map(|res| res.response.hash.to_string())
-        .join(", ");
-
+    tx_hashes: &[TxHash],
+) -> Result<Vec<IbcEvent>, Error> {
     info!(
         id = %chain_id,
-        "wait_for_block_commits: waiting for commit of tx hashes(s) {}",
-        hashes
+        "wait_for_block_commits: waiting for commit of tx hashes(s) {:?}",
+        tx_hashes
     );
 
-    loop {
-        let elapsed = start_time.elapsed();
-
-        if all_tx_results_found(tx_sync_results) {
-            trace!(
-                id = %chain_id,
-                "wait_for_block_commits: retrieved {} tx results after {}ms",
-                tx_sync_results.len(),
-                elapsed.as_millis(),
-            );
-
-            return Ok(());
-        } else if &elapsed > rpc_timeout {
-            return Err(Error::tx_no_confirmation());
-        } else {
-            thread::sleep(WAIT_BACKOFF);
+    let responses = wait_tx_hashes(rpc_client, rpc_address, rpc_timeout, tx_hashes).await?;
 
-            for tx_sync_result in tx_sync_results.iter_mut() {
-                // ignore error
-                let _ =
-                    update_tx_sync_result(chain_id, rpc_client, rpc_address, tx_sync_result).await;
-            }
+    let mut all_events = Vec::new();
+
+    for response in responses.into_iter() {
+        for tx_response in response.txs.into_iter() {
+            let mut events = all_ibc_events_from_tx_search_response(chain_id, tx_response);
+            all_events.append(&mut events);
         }
     }
+
+    Ok(all_events)
 }
 
-async fn update_tx_sync_result(
-    chain_id: &ChainId,
+pub async fn wait_tx_hashes(
     rpc_client: &HttpClient,
     rpc_address: &Url,
-    tx_sync_result: &mut TxSyncResult,
+    timeout: &Duration,
+    tx_hashes: &[TxHash],
+) -> Result<Vec<TxSearchResponse>, Error> {
+    let mut join_handles = Vec::new();
+
+    for tx_hash in tx_hashes.iter() {
+        let rpc_client = rpc_client.clone();
+        let rpc_address = rpc_address.clone();
+        let timeout = *timeout;
+        let tx_hash = *tx_hash;
+
+        let handle = task::spawn(async move {
+            wait_tx_hash(&rpc_client, &rpc_address, &timeout, &tx_hash).await
+        });
+        join_handles.push(handle);
+    }
+
+    let mut responses = Vec::new();
+
+    for handle in join_handles.into_iter() {
+        let response = handle.await.map_err(Error::join)??;
+        responses.push(response);
+    }
```

Comment:
It's more efficient to use `FuturesOrdered`.

Reply:
Thanks for the suggestions! Loops are easier to write during quick hacks, though I have found the time to refactor them to use iterators and streams.